sound/soc/sof/sof-audio.c

```diff
@@ -107,6 +107,17 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 	if (!swidget->private)
 		return 0;
 
+	/*
+	 * PCM streams can be freed without being restarted after system resume. Prevent
+	 * the use_count from being decremented below 0 for dynamic widgets and below 1
+	 * for static widgets in such scenarios.
+	 */
```

Comment:
Sorry, I didn't get the logic. How do we know the PCM stream is freed without being restarted? And why we should prevent the use_count from being decremented? Will this prevent widget from being freed in normal cases?

Reply:
this is what pulse audio does when the system is suspended during probe. It frees the stream that it opened before suspend without ever restarting them. If the use_count is 0 (or 1 for static widgets), we should not decrement it. The only reason you'd ever hit this is if you have a hw_free without hw_params and in that case, we should do nothing.

Reply:
My point is that when will we free the widget? I assume that `use_count` will be decreased when `sof_widget_free` is called for any reason. So ideally the use_count will be 4->3->2->1->0 and we will free the widget when it is 0. But with this change, we will do nothing when the use_count is 0 (or 1 for static widgets). Let's take static widgets for example, the use_count will be 4->3->2->1->1->1..., right? Maybe I am wrong but I am confused here.

Reply:
For static widgets, you'd never free the widget and thats why they are called static. For dynamic widgets, you will free them when the current use_count is 1 when sof_widget_free() is called

Reply:
> For static widgets, you'd never free the widget and thats why they are called static. For dynamic widgets, you will free them when the current use_count is 1 when sof_widget_free() is called

Thanks @ranj063 for the explanation. It answered my question.